### PR TITLE
Data callbacks

### PR DESCRIPTION
**Pull request by xloem**
This a separate PR for my work on #443 .  More description at https://github.com/ITotalJustice/cpr/pull/1 .  Let me know if you'd like a rebase.

I'm realizing that my removal of ITotalJustice's custom structs makes it more difficult to add functionality to fill the Response fields simultaneous with using the callback functions.  That's unfortunate, but this is what I have implemented at this time.

### CODE REVIEW

File: cpr/session.cpp [hunk 1]
```diff
@@ -41,6 +41,11 @@ class Session::Impl {
     void SetCookies(const Cookies& cookies);
     void SetBody(Body&& body);
     void SetBody(const Body& body);
+    void SetReadCallback(const ReadCallback& read);
+    void SetHeaderCallback(const HeaderCallback& header);
+    void SetWriteCallback(const WriteCallback& write);
+    void SetProgressCallback(const ProgressCallback& parameters);
```
Comment (COM8) [inline cpr/session.cpp:47]:
```suggestion
    void SetProgressCallback(const ProgressCallback& progress);
```

File: cpr/session.cpp [hunk 2]
```diff
@@ -330,6 +342,65 @@ void Session::Impl::SetBody(const Body& body) {
...
+        curl_easy_setopt(curl, CURLOPT_INFILESIZE_LARGE,
+                         static_cast<curl_off_t>(read.size));
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE_LARGE,
+                         static_cast<curl_off_t>(read.size));
+        curl_easy_setopt(curl, CURLOPT_READFUNCTION, cpr::util::readUserFunction);
+        curl_easy_setopt(curl, CURLOPT_READDATA, &readcb_);
+        if (read.size == -1) {
+            SetHeader({{"Transfer-Encoding", "chunked"}});
+        }
+        readcb_ = read;
```
Comment (COM8) [inline cpr/session.cpp:357]:
```suggestion
    if (curl) {
        readcb_ = read;
        curl_easy_setopt(curl, CURLOPT_INFILESIZE_LARGE,
                         static_cast<curl_off_t>(read.size));
        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE_LARGE,
                         static_cast<curl_off_t>(read.size));
        curl_easy_setopt(curl, CURLOPT_READFUNCTION, cpr::util::readUserFunction);
        curl_easy_setopt(curl, CURLOPT_READDATA, &readcb_);
        if (read.size == -1) {
            SetHeader({{"Transfer-Encoding", "chunked"}});
        }
```
Better readability.

File: cpr/session.cpp [hunk 3]
```diff
@@ -330,6 +342,65 @@ void Session::Impl::SetBody(const Body& body) {
...
+    }
+}
+
+void Session::Impl::SetHeaderCallback(const HeaderCallback& header) {
+    CURL* curl = curl_->handle;
+    if (curl) {
+        curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION , cpr::util::headerUserFunction);
+        curl_easy_setopt(curl, CURLOPT_HEADERDATA, &headercb_);
+        headercb_ = header;
```
Comment (COM8) [inline cpr/session.cpp:366]:
```suggestion
        headercb_ = header;
        curl_easy_setopt(curl, CURLOPT_HEADERDATA, &headercb_);
```

File: cpr/session.cpp [hunk 4]
```diff
@@ -330,6 +342,65 @@ void Session::Impl::SetBody(const Body& body) {
...
+    }
+}
+
+void Session::Impl::SetWriteCallback(const WriteCallback& write) {
+    CURL* curl = curl_->handle;
+    if (curl) {
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, cpr::util::writeUserFunction);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &writecb_);
+        writecb_ = write;
```
Comment (COM8) [inline cpr/session.cpp:375]:
```suggestion
        writecb_ = write;
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &writecb_);
```

File: cpr/session.cpp [hunk 5]
```diff
@@ -330,6 +342,65 @@ void Session::Impl::SetBody(const Body& body) {
...
+        #if LIBCURL_VERSION_NUM < 0x072000
+        curl_easy_setopt(curl, CURLOPT_PROGRESSFUNCTION, cpr::util::progressUserFunction);
+        curl_easy_setopt(curl, CURLOPT_PROGRESSDATA, &progresscb_);
+        #else
+        curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, cpr::util::progressUserFunction);
+        curl_easy_setopt(curl, CURLOPT_XFERINFODATA, &progresscb_);
+        #endif
+        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 0L);
+        progresscb_ = progress;
+    }
```
Comment (COM8) [inline cpr/session.cpp:391]:
```suggestion
    if (curl) {
        progresscb_ = progress;
        #if LIBCURL_VERSION_NUM < 0x072000
        curl_easy_setopt(curl, CURLOPT_PROGRESSFUNCTION, cpr::util::progressUserFunction);
        curl_easy_setopt(curl, CURLOPT_PROGRESSDATA, &progresscb_);
        #else
        curl_easy_setopt(curl, CURLOPT_XFERINFOFUNCTION, cpr::util::progressUserFunction);
        curl_easy_setopt(curl, CURLOPT_XFERINFODATA, &progresscb_);
        #endif
        curl_easy_setopt(curl, CURLOPT_NOPROGRESS, 0L);
    }
```

File: cpr/session.cpp [hunk 6]
```diff
@@ -330,6 +342,65 @@ void Session::Impl::SetBody(const Body& body) {
...
+}
+
+void Session::Impl::SetDebugCallback(const DebugCallback& debug) {
+    CURL* curl = curl_->handle;
+    if (curl) {
+        curl_easy_setopt(curl, CURLOPT_DEBUGFUNCTION, cpr::util::debugUserFunction);
+        curl_easy_setopt(curl, CURLOPT_DEBUGDATA, &debugcb_);
+        curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+        debugcb_ = debug;
```
Comment (COM8) [inline cpr/session.cpp:400]:
```suggestion
        debugcb_ = debug;
        curl_easy_setopt(curl, CURLOPT_DEBUGDATA, &debugcb_);
        curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
```

**Review by COM8 — CHANGES_REQUESTED** [review]
Great work. Thanks!
Only a few minor changes.

**Comment by COM8** [comment]
The following needs to be fixed for the windows CI to run:
```
D:\a\cpr\cpr\include\cpr/callback.h(14,93): error C2614: 'cpr::ReadCallback': illegal member initialization: 'size' is not a base or member [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
```
Source: https://github.com/whoshuu/cpr/pull/446/checks?check_run_id=1144990491

**Comment by xloem** [comment]
It looks like the CI failure is from windows not having `ssize_t`.  I (or somebody) will have to evaluate the context of use again and either pick a different type or accommodate windows.  There's probably a type already used similarly in cpr in other places.  Makes sense to add the clarity suggestions at the same time.

**Comment by COM8** [comment]
Windows does not know the `ssize_t` type. I will push a quick fix for this.

**Review by COM8 — APPROVED** [review]

**Comment by xloem** [comment]
Thanks for pulling in all this stuff com8, we can all celebrate.

**Pull request merged**